cmd/swarm/clientaccountmanager.go

```diff
@@ -0,0 +1,103 @@
+// Copyright 2016 The go-ethereum Authors
+// This file is part of go-ethereum.
+//
+// go-ethereum is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// go-ethereum is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with go-ethereum. If not, see <http://www.gnu.org/licenses/>.
+
+/*
+//This module implements the bare minimum to allow client commands to access the keystore
+*/
+package main
+
+import (
+	"crypto/ecdsa"
+	"fmt"
+	"io/ioutil"
+	"os"
+
+	"github.com/ethereum/go-ethereum/accounts"
+	"github.com/ethereum/go-ethereum/accounts/keystore"
+	"github.com/ethereum/go-ethereum/accounts/usbwallet"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/node"
+	cli "gopkg.in/urfave/cli.v1"
+
+	"github.com/ethereum/go-ethereum/cmd/utils"
+)
+
+type ClientAccountManagerProvider struct {
+	manager *accounts.Manager
+}
+
+func (c *ClientAccountManagerProvider) AccountManager() *accounts.Manager {
+	return c.manager
+}
+
+func getClientAccount(ctx *cli.Context) *ecdsa.PrivateKey {
```

Comment:
identity management and crypto should be solved properly for both MRU signing and Access Control. 
we need to use clef or a javascript directy in the browser for non-cli.

for cli we also need to share code @justelad will comment

Reply:
This code allows the user to access keys stored in the Ethereum keystore for signing MRUs, for his/her convenience.

Otherwise, the user would have to provide a key in a different way. This was not meant to be a final implementation, but rather a way to quickly give access to the feature without much work for the user.

If this is an issue, I can remove `clientaccountmanager.go` altogether and ask the user to point the cli to a private key on disk, or environment variable, etc, e.g.

`swarm resource create 300 --name "myresource" --key "/home/jm/mykey.txt"`

Reply:
> but rather a way to quickly give access to the feature without much work for the user

Yes, it should be quick and dirty, and having to separately give access to (something that gives access to) the key is too complicated in my opinion, at least as the only option.

Reply:
Alternatively, and part of the current implementation, the user can point `--bzzaccount` to a private key file path (encrypted or unencrypted), instead of specifying an address.
If the private key is encrypted, the user can also use `--password`, otherwise they'll get a prompt.
Thus, the following schemes are possible:
* `swarm --bzzaccount="/home/jm/key.txt" resource create 300 --name "myresource"`
* `swarm --bzzaccount="/home/jm/key.txt" --password "xyz1234" resource create 300 --name "myresource"`
and the regular one that looks for the key in the keystore:
* `swarm --bzzaccount="0x77b68bf07bdf9986084affaca2eace59700bf644" --password "xyz1234" resource create 300 --name`

This is not part of this PR, it is already the way it works for starting a node.

Reply:
btw, we just had to do the same thing (more or less) for the access control tries.

you can see the diff [here](https://github.com/ethersphere/go-ethereum/compare/master...ethersphere:access-control?expand=1)
look for `func doPKNew` in `access.go`.

Reply:
>  i didn't quite get why you need to shadow the ethereum code for managing keystores? 

@justelad I tried, the problem is that that account management code is tightly coupled to execute when the node is started, and I don't want to start a node. I also didn't want to touch core Ethereum packages.

Another option is to make `node/config.go` `makeAccountManager()` public and call it. But that meant touching a core Ethereum package, and I am not sure if we can do that within Swarm's scope.

Reply:
> btw, we just had to do the same thing (more or less) for the access control tries.

Yes, I see, however you start a node there, and for signing a MRU I shouldn't need to start a node.

Reply:
@justelad @zelig @nolash . can we land this as is, and mabye make a PR to Ethereum to change `node/config.go` so we make `makeAccountManager()` public? Once that PR is in, I can refactor this code to use it instead of our copy.

Thanks

Reply:
Look closer - we don’t really spin up a node there, we just use the builtin functionality to construct one and get the keystore data from it. For the time being no need to expose anything from the node package IMHO. I’d really be happy to see this file go, as the concept of changing something in the main geth ethereum keystore/account management logic will be very hard to swallow and we can’t count on it happening.

Reply:
Hi @justelad. I took a closer look and I tried to implement your method. However, I ran into a problem that you will probably also have.

In your function `doPKNew`, line 166 you have the following code:
```go
	}
	initSwarmNode(bzzconfig, stack, ctx)
	privateKey := getAccount(bzzconfig.BzzAccount, ctx, stack)
```
This works, however, `initSwarmNode()` also calls `getAccount()` inside. Thus, you're calling `getAccount()` twice. The issue with this is that if the account is encrypted you get a password prompt twice and that is *very* annoying.

I distilled the essence of this into this function maybe we can both use:

```go
// getPrivKey returns the private key of the specified bzzaccount
// Used only by client commands, such as `resource`
func getPrivKey(ctx *cli.Context) *ecdsa.PrivateKey {
	// booting up the swarm node just as we do in bzzd action
	bzzconfig, err := buildConfig(ctx)
	if err != nil {
		utils.Fatalf("unable to configure swarm: %v", err)
	}
	cfg := defaultNodeConfig
	if _, err := os.Stat(bzzconfig.Path); err == nil {
		cfg.DataDir = bzzconfig.Path
	}
	utils.SetNodeConfig(ctx, &cfg)
	stack, err := node.New(&cfg)
	if err != nil {
		utils.Fatalf("can't create node: %v", err)
	}
//	initSwarmNode(bzzconfig, stack, ctx)
	return getAccount(bzzconfig.BzzAccount, ctx, stack)
}
```
Notice that I removed the `initSwarmNode()` call, compared with your code in `doPKNew`

I have update this PR and now the infamous `clientaccountmanager.go` is gone!